OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
+
+				var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
+				if (NextActivity != null)
+					return Util.SequenceActivities(NextActivity, new Wait(randFrames), this);
 				else
-				{
-					// Get out of the way if we are:
-					harv.UnblockRefinery(self);
-					var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
-					if (NextActivity != null)
-						return Util.SequenceActivities(NextActivity, new Wait(randFrames), new FindResources());
-					else
-						return Util.SequenceActivities(new Wait(randFrames), new FindResources());
-				}
+					return Util.SequenceActivities(new Wait(randFrames), this);
 			}
 
-			// Attempt to claim a resource as ours:
+			// Attempt to claim a resource as ours
 			if (territory != null)
 			{
-				if (!territory.ClaimResource(self, path[0]))
-					return Util.SequenceActivities(new Wait(25), new FindResources());
+				if (!territory.ClaimResource(self, closestHarvestablePosition.Value))
+					return Util.SequenceActivities(new Wait(25), this);
 			}
 
 			// If not given a direct order, assume ordered to the first resource location we find:
-			if (harv.LastOrderLocation == null)
-				harv.LastOrderLocation = path[0];
+			if (!harv.LastOrderLocation.HasValue)
+				harv.LastOrderLocation = closestHarvestablePosition;
 
-			self.SetTargetLine(Target.FromCell(self.World, path[0]), Color.Red, false);
+			self.SetTargetLine(Target.FromCell(self.World, closestHarvestablePosition.Value), Color.Red, false);
 
 			var notify = self.TraitsImplementing<INotifyHarvesterAction>();
-			var next = new FindResources();
+			var next = this;
 			foreach (var n in notify)
-				n.MovingToResources(self, path[0], next);
+				n.MovingToResources(self, closestHarvestablePosition.Value, next);
 
-			return Util.SequenceActivities(mobile.MoveTo(path[0], 1), new HarvestResource(), new FindResources());
+			return Util.SequenceActivities(mobile.MoveTo(closestHarvestablePosition.Value, 1), new HarvestResource(), next);
 		}
 
-		// Diagonal distance heuristic
-		static int EstimateDistance(CPos here, CPos destination)
+		bool IsHarvestable(Actor self, CPos pos)
 		{
-			var diag = Math.Min(Math.Abs(here.X - destination.X), Math.Abs(here.Y - destination.Y));
-			var straight = Math.Abs(here.X - destination.X) + Math.Abs(here.Y - destination.Y);
+			var resType = resLayer.GetResource(pos);
+			if (resType == null)
+				return false;
+
+			// Can the harvester collect this kind of resource?
+			if (!harvInfo.Resources.Contains(resType.Info.Name))
+				return false;
+
+			if (territory != null)
+			{
+				// Another harvester has claimed this resource:
+				ResourceClaim claim;
+				if (territory.IsClaimedByAnyoneElse(self, pos, out claim))
+					return false;
+			}
+
+			return true;
+		}
+
+		/// <summary>
+		/// Finds the closest harvestable pos between the current position of the harvester
+		/// and the last order location
+		/// </summary>
+		CPos? ClosestHarvestablePos(Actor self)
```

Comment:
Something in here regressed, so that (some) of the harvesters in the intervention mission stop working after dropping off their first load..